src/main/java/ru/yandex/practicum/filmorate/service/UserService.java

```diff
@@ -51,11 +51,11 @@ public void requestFriendship(Long userId, Long friendToAddId) throws UserNotFou
         userStorage.requestFriendship(userId, friendToAddId);
     }
 
-    public void deleteFriend(Long userId, Long friendToDellId) throws UserNotFoundException, FilmNotFoundException {
+    public void deleteFriend(Long userId, Long friendToDellId) throws UserNotFoundException {
         userStorage.deleteFriendFromUser(userId, friendToDellId);
     }
 
-    public Collection<Long> getMutualFriendsIds(Long user1Id, Long user2Id) throws UserNotFoundException, FilmNotFoundException, MpaNotFoundException, GenreNotFoundException, DirectorNotFoundException {
+    public Collection<Long> getMutualFriendsIds(Long user1Id, Long user2Id) throws UserNotFoundException {
         long userIdWithLessFriends = Long.min(getFriendsNum(user1Id), getFriendsNum(user2Id));
```

Comment:
This logic has O(n) unnecessary queries to the db. Consider the following:

```
User firstUserFriends = userStorage.getById(firstUserId).getFriends().keySet();
User secondUserFriends = userStorage.getById(secondUserId).getFriends().keySet();

List<Long> commonFriends = firstUserFriends.size() >= secondUserFriends.size() ? firstUserFriends : secondUserFriends;
commonFriends.retainAll(firstUserFriends);
commonFriends.retainAll(secondUserFriends); //awkward, can be optimized.

return commonFriends;
```

Alternatively and potentially quicker would be to make an inner join on DB side.

Reply:
Thank you! Missed this legacy from version with no DB.